Tests/Adapters/Incoming/CreateTaskTest.php

```diff
@@ -0,0 +1,27 @@
+<?php
+
+namespace Tests\Adapters\Incoming;
+
+use Application\Task\TaskService;
+use Infrastructure\Terminal\Commands\CreateTask;
+use PHPUnit\Framework\TestCase;
+use Symfony\Component\Console\Tester\CommandTester;
+
+class CreateTaskTest extends TestCase
+{
+    public function test_a_task_can_be_created_through_the_command_line(): void
```

Comment:
This looks like a perfect candidate for a full integration test. Boot the testing container and provide actual input and verify the output. It'll verify the full implementation:
* Input argument validation
* Task creation
* Output message verification

Especially in small applications like this, I believe mocking is unnecessary. The `TaskRepositoryInMemory` is a 100% valid replacement for the actual SQL implementation, so I'd use it 😄 .

Reply:
I agree, there's currently a missing an "end-to-end" test in the application. But I'd argue that I would not put such a test in this test, as this is a more isolated "adapter test". To my understanding, an adapter test does not need to know what happens in our core code, and should only be concerned with the infrastructure/framework.

With a full test, indeed we don't have to mock anything. But now I'm thinking about your statement that mocking in general is unnecessary in smaller applications.. now I'm not sure anymore 😅.

I would put some more time in exploring this, but for now my head is calling me to go early to bed. Would love to continue this discussion tomorrow though, I don't think it would be a very long one either :)

Reply:
Of course, anything in programming is debatable 🙂 .

For me it's simple, if I write a full implementation test, I don't have to write an adapter test. And the test will be super simple and readable with your current setup.

```php
final class CreateTaskTest extends TestCase
{
    public function test_it_can_create_a_task(): void
    {
        $taskService = new TaskService(new TaskRepositoryInMemory());
        
        $command = new CreateTask($taskService);

        $commandTester = new CommandTester($command);
        
        $response = $commandTester->execute(['title' => 'Foo']);
        
        // Assert the repository contains a task with title 'Foo';
        // Assert the output contains 'Task created!';
    }
}
```

This is less lines of codes than your current test, and actually hits all services and repositories involved. I'd always choose this over mocking because it's easier to read, it's complete, and validates expected outcome instead of verifying the implementation.

By booting the container I mean that you could choose to fetch the `CreateTask` console command from the container:

```php
final class TestServiceContainerWithDatabase extends ServiceContainer
{
    public function createTaskCliCommand(): CreateTask
    {
        return new CreateTask(new TaskRepositoryInMemory());
    }
}
```

You created a service container by hand (which is fine for a small app like this). You just have to new up the container in your test to "boot" it.